File: frontend/src/types/repository.ts
export type RepositoryVisibility = 'public' | 'private' | 'government' | 'affiliates_only'
export type RepositoryType = 'jurisdiction' | 'policy_area' | 'budget'

export interface RepositoryOwner {
  id: number
  username: string
  full_name?: string | null
}

export interface Repository {
  id: number
  name: string
  slug: string
  description?: string | null
  type: RepositoryType
  visibility: RepositoryVisibility
  owner_id?: number | null
  owner?: RepositoryOwner | null
  jurisdiction_name?: string | null
  jurisdiction_type?: string | null
  allow_public_proposals: boolean
  allow_public_voting: boolean
  require_verification_for_voting: boolean
  quorum_percentage: number
  voting_period_days: number
  min_signatures_for_voting: number
  proposals_count: number
  issues_count: number
  contributors_count: number
  is_active: boolean
  is_archived: boolean
  created_at: string
  updated_at: string
}

export interface Commit {
  id: number
  repository_id: number
  author_id: number
  message: string
  content: string
  previous_commit_id?: number
  created_at: string
}

export interface File {
  id: number
  repository_id: number
  name: string
  content: string
  current_commit_id: number
  created_at: string
  updated_at: string
}

File: backend/app/schemas/repository.py
from pydantic import BaseModel, validator
from typing import Optional, List
from datetime import datetime
from app.models.repository import RepositoryType, RepositoryVisibility

class UserOwner(BaseModel):
    """Informações básicas do usuário proprietário do repositório"""
    id: int
    username: str
    full_name: Optional[str] = None
    
    class Config:
        from_attributes = True

class RepositoryBase(BaseModel):
    name: str
    description: Optional[str] = None
    type: RepositoryType
    jurisdiction_name: Optional[str] = None
    jurisdiction_type: Optional[str] = None
    
    @validator('name')
    def validate_name(cls, v):
        if not v or len(v) < 3:
            raise ValueError('Repository name must be at least 3 characters long')
        return v

class RepositoryCreate(RepositoryBase):
    visibility: RepositoryVisibility = RepositoryVisibility.PUBLIC
    allow_public_proposals: bool = True
    allow_public_voting: bool = True
    require_verification_for_voting: bool = True
    quorum_percentage: int = 10
    voting_period_days: int = 7
    min_signatures_for_voting: int = 500

    @validator("visibility", pre=True)
    def normalize_visibility(cls, v):
        if v is None:
            return RepositoryVisibility.PUBLIC
        if isinstance(v, RepositoryVisibility):
            return v
        return RepositoryVisibility(str(v).lower())


class RepositoryForkRequest(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    visibility: Optional[RepositoryVisibility] = None
    jurisdiction_name: Optional[str] = None
    jurisdiction_type: Optional[str] = None

    @validator("visibility", pre=True, always=False)
    def normalize_visibility(cls, v):
        if v is None:
            return None
        if isinstance(v, RepositoryVisibility):
            return v
        return RepositoryVisibility(str(v).lower())

class RepositoryUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    visibility: Optional[RepositoryVisibility] = None
    allow_public_proposals: Optional[bool] = None
    allow_public_voting: Optional[bool] = None
    require_verification_for_voting: Optional[bool] = None
    quorum_percentage: Optional[int] = None
    voting_period_days: Optional[int] = None
    min_signatures_for_voting: Optional[int] = None

    @validator("visibility", pre=True, always=False)
    def normalize_visibility(cls, v):
        if v is None:
            return None
        if isinstance(v, RepositoryVisibility):
            return v
        return RepositoryVisibility(str(v).lower())

class RepositoryInDB(RepositoryBase):
    id: int
    slug: str
    visibility: RepositoryVisibility
    is_active: bool
    is_archived: bool
    proposals_count: int
    issues_count: int
    contributors_count: int
    owner_id: Optional[int] = None
    owner: Optional[UserOwner] = None
    is_fork: bool
    forked_from_id: Optional[int] = None
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True

class Repository(RepositoryInDB):
    pass

class RepositoryPublic(BaseModel):
    id: int
    name: str
    slug: str
    description: Optional[str]
    type: RepositoryType
    jurisdiction_name: Optional[str]
    jurisdiction_type: Optional[str]
    visibility: RepositoryVisibility
    proposals_count: int
    issues_count: int
    contributors_count: int
    owner_id: Optional[int] = None
    owner: Optional[UserOwner] = None
    created_at: datetime
    
    class Config:
        from_attributes = True
